### PR TITLE
Implemented Inventory Strategy

### PR DESCRIPTION
**Pull request by lilmont**
An inventory strategy which returns products wither in FIFO or LIFO manner.

### CODE REVIEW

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    private T? _current;
+    private bool _isDisposed;
+    public FIFOEnumerator(List<T> list)
+    {
+        _list = list;
+        _index = 0;
+        _current = default;
+        _isDisposed = false;
+    }
+    public T Current
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:18]: please follow 1 line break for each method.

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace InventoryStrategy.Enumerators;
+
+public class FIFOEnumerator<T> : IEnumerator<T>
+{
+    private readonly List<T> _list;
+    private int _index;
+    private T? _current;
+    private bool _isDisposed;
+    public FIFOEnumerator(List<T> list)
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:11]: please follow 1 line break for each method.

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+﻿using System.Collections;
+
+namespace InventoryStrategy.Enumerators;
+
+public class FIFOEnumerator<T> : IEnumerator<T>
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:5]: please write summary for you class.

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+﻿using InventoryStrategy.Enumerators;
+using InventoryStrategy.Models.Enums;
+using System.Collections;
+
+namespace InventoryStrategy;
+
+public class InventoryWrapper<T> : IEnumerable<T>
+{
+    private InventoryManagementApproach _managementApproach;
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:9]: this field must be readonly. if you fill it in your constructor.

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    private List<T> _list;
+    public InventoryWrapper(InventoryManagementApproach managementApproach)
+    {
+        _managementApproach = managementApproach;
+        _list = new List<T>();
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:14]: use collection Expression in c# 12.

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    public void Add(T item)
+    {
+        _list.Add(item);
+    }
+    public IEnumerator<T> GetEnumerator()
+    {
+        if (_managementApproach is InventoryManagementApproach.LastInFirstOut)
+        {
+            LIFOEnumerator<T> enumerator = new(_list);
+            while (enumerator.MoveNext())
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:25]: DRY

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                yield return enumerator.Current;
+        }
+        if (_managementApproach is InventoryManagementApproach.FirstInFirstOut)
+        {
+            FIFOEnumerator<T> enumerator = new(_list);
+            while (enumerator.MoveNext())
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:31]: DRY. please refactor it. 
you are writing the same code in different if.

File: sources/Niloo Mont/InventoryStrategy/Models/Enums/InventoryManagementApproach.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿namespace InventoryStrategy.Models.Enums;
+
+public enum InventoryManagementApproach
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Models/Enums/InventoryManagementApproach.cs:3]: inherit of byte. (micro-optimization)
in this case, it is not important, but if you want to map this enum in your DB you should inherit from byte.

File: sources/Niloo Mont/InventoryStrategy/Models/Phone.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿namespace InventoryStrategy.Models;
+
+public class Phone : Product
+{
+    public Phone(int code, string name) : base(code, name)
+    {
+
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Models/Phone.cs:7]: remove use-less break line.

File: sources/Niloo Mont/InventoryStrategy/Models/Phone.cs [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿namespace InventoryStrategy.Models;
+
+public class Phone : Product
+{
+    public Phone(int code, string name) : base(code, name)
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Models/Phone.cs:5]: use primary constructor.

File: sources/Niloo Mont/InventoryStrategy/Models/Product.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿namespace InventoryStrategy.Models;
+
+public class Product
+{
+    public int Code { get; set; }
+
+    public string Name { get; set; } = null!;
+
+    public Product(int code, string name)
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Models/Product.cs:9]: Primary constructor.

File: sources/Niloo Mont/InventoryStrategy/Models/TV.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+﻿namespace InventoryStrategy.Models;
+
+public class TV : Product
+{
+    public TV(int code, string name) : base(code, name)
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Models/TV.cs:5]: again Primary constructor.

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]
your code was perfect, but I want you to be better than now. so I reject your code to improve it. 
your faults were not about functionality, all of them were about clean code, and principals.

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs (hunk 5 above)
Comment (lilmont) [reply sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:14]: I tried changing this line to _list = []; but the compiler doesn't recognize this pattern. How can I resolve it?

File: sources/Niloo Mont/InventoryStrategy/Models/Phone.cs (hunk 10 above)
Comment (lilmont) [reply sources/Niloo Mont/InventoryStrategy/Models/Phone.cs:5]: I have the same problem I had with collection Expression. I would appreciate if you could help me with that.

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs (hunk 3 above)
Comment (mohammadKarimi) [reply sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:7]: for summary: please write /// it makes you a standard summary.

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+    public void Dispose()
+    {
+        if (!_isDisposed)
+        {
+            _list.Clear();
+            _isDisposed = true;
+        }
+    }
+    public bool MoveNext()
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:37]: line break.

File: sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        List<T> localList = _list;
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/FIFOEnumerator.cs:39]: why you are assigning to localList?

File: sources/Niloo Mont/InventoryStrategy/Enumerators/LIFOEnumerator.cs [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        List<T> localList = _list;
+        if (_index >= 0)
+        {
+            _current = localList[_index];
+            _index--;
+            return true;
+        }
+        return false;
+    }
+    public void Reset()
```
Comment (mohammadKarimi) [inline sources/Niloo Mont/InventoryStrategy/Enumerators/LIFOEnumerator.cs:48]: line break.

File: sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs (hunk 5 above)
Comment (mohammadKarimi) [reply sources/Niloo Mont/InventoryStrategy/InventoryWrapper.cs:16]: it is not important. it is ok, I am sending you some tips to use.

File: sources/Niloo Mont/InventoryStrategy/Models/Phone.cs (hunk 10 above)
Comment (mohammadKarimi) [reply sources/Niloo Mont/InventoryStrategy/Models/Phone.cs:5]: are you using .net 8 ?

**Review by mohammadKarimi — APPROVED** [review]

**Pull request merged**